anndata/_core/merge.py

```diff
@@ -479,6 +514,10 @@ def gen_inner_reindexers(els, new_index, axis: Literal[0, 1] = 0):
             lambda x, y: x.intersection(y), (df_indices(el) for el in els)
         )
         reindexers = [Reindexer(df_indices(el), common_ind) for el in els]
+    elif all(isinstance(el, AwkArray) for el in els if not_missing(el)):
+        # do not reindex awkward arrays
+        # TODO unintended behaviour?
+        reindexers = [lambda *args, **kwargs: args[0] for _ in els]
```

Comment:
What do these reindexers do during merge and what happens if we don't reindex?

Reply:
Are you asking what `Reindexer`s do generally, or specifically what the ones defined here do?

Reply:
these ones specifically. 

I *think* they are relevant for subsetting columns if `obsm`/`varm` is a data frame? I that case I think it would indeed be the right thing to do nothing for awkward arrays, as the second axis is not aligned to anything and could be of variable length.

Reply:
> I think they are relevant for subsetting columns if obsm/varm is a data frame?

Yes. It's for taking the intersection of the available columns, so we are just concatenating those for an inner join.

For non-labelled arrays (ndarrays and sparse) we take minimum of the column axis size.

> I think it would indeed be the right thing to do nothing for awkward arrays, as the second axis is not aligned to anything and could be of variable length.

That's where I get a little confused, since it depends on the dimensionality of the awkward array.

I think what inner join would be, would be to take the intersection of all axes apart from the one we're concatenating along.

For example, lets say we have:

```python
awk_a = ak.Array([
    {"a": [1, 2, 3], "b": [1, 2]},
    {"a": [4, 5], "b": [3, 4]},
    {"a": [6], "b": [5]},
])


awk_b = ak.Array([
    {"a": [1, 2, 3]},
    {"a": [4, 5]},
    {"a": [6]},
])
```

For an inner join I think we take the intersection of keys from the `Record` dimension, dropping the `"b"` key from the record type.

For an outer join, I think I would expect an array where `b` is an optional type.

However: I see awkard array makes a union type here – so I'm not too sure what's right here.

Reply:
This would only work for a `RecordType`, what would you do for a `ListType`? 

```python
awk_a  = ak.Array([
    [1,2,3],
    [1,2,3,4],
])

awk_b = ak.Array([
  [],
  [1,2]
  [1,2,3,4,5,6]
])
```

Also, even when you are using a RecordType, unlike for a dataframe, there's no need that every record has the same columns. So I would just keep it simple and just concatenate them.

Reply:
> what would you do for a ListType?

A record type has a fixed set of keys, while a list type is inherently unsized. I'd consider `ListTypes` equivalent, while `RecordType`s are only aligned if the keys are the same.

I sorta see it as we have already defined `inner` (intersection) and `outer` (union) for Record and Regular dimensions. I think they don't really have a meaning for `List` dimensions, so don't do anything.

> Also, even when you are using a RecordType, unlike for a dataframe, there's no need that every record has the same columns. So I would keep it simple and just concatenate them.

This is one option. That `join` has no meaning for awkward arrays.

But, I think there is quite a lot of usefulness to the `inner` join. We are dropping fields which aren't represented for all samples. Do you actually want to always keep those around?

I think it also becomes confusing in the case where one of the dimensions is Regular. Now we can have different `Regular` dimensions per row. What does this mean for concatenating values in `X`?

Reply:
I'm a bit afraid that this ends up being a rabbit hole, as the arrays can be arbitrarily nested. If you have a record type that's embedded in a variable-length list type in the 6th dimension, would you still want to attempt an inner/outer join? There might also be nasty `UnionType` combinations (technically it is possible to have both records and lists in the same `ListType`). 

Could we get away with not performing a merge for now, and implement a merging strategy later (if requested) in a backwards-compatible manner?

Reply:
> Could we get away with not performing a merge for now, and implement a merging strategy later (if requested) in a backwards-compatible manner?

I'm not sure how we could do this in a backwards compatible way, since the current logic isn't a subset of the intersection. If you want to punt on this I think we can say awkward array support is experimental and the concatenation logic will change in the future.

---------

I don't think the arbitrarily nested types would be too difficult to work with. We could just iterate through the levels applying the same logic at each level.

`UnionType`s however, I'm not so sure how to deal with.

Reply:
I suggest we go with



 >> Also, even when you are using a RecordType, unlike for a dataframe, there's no need that every record has the same columns. So I would keep it simple and just concatenate them.
>
> This is one option. That join has no meaning for awkward arrays.

for now. 

We can create an issue to track adding support for inner joins of awkward arrays (or at least various special cases thereof, such as RecordTypes in the second dimension. Since we decided on releasing this as experimental, I wouldn't worry to much about backwards compatibility.